# ==========================================
# Exploratory Data Analysis (EDA)
# ==========================================

## Objective

The objective of this notebook is to explore the cleaned veterinary radiology dataset in order to identify meaningful patterns, trends and relationships.

The analysis focuses on answering practical business and clinical questions such as:

- Which species are examined most frequently?
- Which breeds are the most common?
- What is the distribution of patient age and weight?
- How many radiographs are usually performed?
- How often are medical reports generated?
- Which veterinarians refer the largest number of cases?
- How does radiology workload change throughout the year?

The insights obtained in this notebook will later support the development of dashboards and business intelligence reports.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.io as pio

pio.renderers.default = "browser"

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
# ==========================================
# LOAD CLEAN DATASET
# ==========================================

radiology = pd.read_csv(
    "../data/processed/radiology_clean.csv",
    parse_dates=["exam_date", "birth_date"]
)

print("Dataset loaded successfully!")

Dataset loaded successfully!


In [3]:
# ==========================================
# FIRST LOOK AT THE DATA
# ==========================================

radiology.head()

,record_id,exam_date,patient_name,species,breed,birth_date,weight_kg,sex,sterilized,owner_name,referring_veterinarian,radiograph_count,medical_report,owner_contact,species_original,breed_original,sex_original,sterilized_original,weight_original,birth_date_original,exam_date_original,radiograph_count_original,medical_report_original,owner_contact_original
0,4315,2022-01-18,LIMA,FELINO,CRIOLLO,2021-09-02,3.0,H,NaN,CAMILO RAMIREZ,LUISA VARGAS,NaN,SI,NaN,FELINO,CRIOLLO,H,NS,3,2021-09-02 00:00:00,2022-01-18 00:00:00,NaN,SI,NO
1,4316,2021-02-19,THORE,CANINO,BULL DOG INGLES,2020-10-01,16.0,M,SI,CAROLINA SUAREZ GARZON,JULIO CESAR DUARTE,NaN,SI,NaN,CANINO,BULL DOG INGLES,M,SI,16,2020-10-01 00:00:00,2021-02-19 00:00:00,NaN,SI,NO
2,4317,2021-02-19,ESTRELLITA,CANINO,BEAGLE,2013-02-01,11.0,H,SI,LORENA RODRIGUEZ,ERWIN RESTREPO,NaN,SI,NaN,CANINO,BEAGLE,H,SI,11,2013-02-01 00:00:00,2021-02-19 00:00:00,NaN,SI,NO
3,4318,2022-02-19,MILU,CANINO,PASTOR ALEMAN,2013-01-01,23.0,H,NO,DUVAN PEÑA,JUAN CAMILO TRUJILLO,NaN,NO,NaN,CANINO,PASTOR ALEMAN,H,NO,23,01.01.2013,2022-02-19 00:00:00,NaN,NO,NO
4,4319,2022-02-21,TAYLER,CANINO,PIT BULL,2020-01-01,28.0,M,NO,LEONARDO CUELLAR,JAIME VEGA,NaN,NO,NaN,CANINO,PIT BULL,M,NO,28,01.01.2020,2022-02-21 00:00:00,NaN,NO,NO


In [4]:
radiology.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 693 entries, 0 to 692
Data columns (total 24 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   record_id                  693 non-null    int64         
 1   exam_date                  689 non-null    datetime64[ns]
 2   patient_name               693 non-null    object        
 3   species                    690 non-null    object        
 4   breed                      686 non-null    object        
 5   birth_date                 533 non-null    datetime64[ns]
 6   weight_kg                  595 non-null    float64       
 7   sex                        683 non-null    object        
 8   sterilized                 595 non-null    object        
 9   owner_name                 691 non-null    object        
 10  referring_veterinarian     688 non-null    object        
 11  radiograph_count           653 non-null    float64       
 12  medical_

In [5]:
radiology.describe(include="all")

,record_id,exam_date,patient_name,species,breed,birth_date,weight_kg,sex,sterilized,owner_name,referring_veterinarian,radiograph_count,medical_report,owner_contact,species_original,breed_original,sex_original,sterilized_original,weight_original,birth_date_original,exam_date_original,radiograph_count_original,medical_report_original,owner_contact_original
count,693.000000,689,693,690,686,533,595.000000,683,595,691,688,653.000000,651,28,692,691,690,681,665,689,690,657,662,681
unique,NaN,NaN,480,3,73,NaN,NaN,2,2,674,195,NaN,2,28,17,109,7,9,182,382,247,8,9,34
top,NaN,NaN,LUNA,CANINO,CRIOLLO,NaN,NaN,M,SI,NS,ERWIN RESTREPO,NaN,NO,3112101288,CANINO,CRIOLLO,M,SI,NS,01.01.2012,2022-11-26 00:00:00,2,NO,NO
freq,NaN,NaN,16,496,256,NaN,NaN,373,363,6,78,NaN,350,1,474,234,369,356,66,18,9,362,345,628
mean,4668.102453,2022-07-30 23:43:16.806966528,NaN,NaN,NaN,2015-10-09 20:26:33.996247552,12.804706,NaN,NaN,NaN,NaN,1.664625,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,460.000000,2021-02-19 00:00:00,NaN,NaN,NaN,1998-01-01 00:00:00,0.300000,NaN,NaN,NaN,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,4487.000000,2022-05-16 00:00:00,NaN,NaN,NaN,2012-02-02 00:00:00,4.000000,NaN,NaN,NaN,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,4679.000000,2022-08-08 00:00:00,NaN,NaN,NaN,2016-01-01 00:00:00,8.600000,NaN,NaN,NaN,NaN,2.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,4854.000000,2022-10-19 00:00:00,NaN,NaN,NaN,2020-02-01 00:00:00,19.100000,NaN,NaN,NaN,NaN,2.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
max,5028.000000,2022-12-31 00:00:00,NaN,NaN,NaN,2022-10-01 00:00:00,50.000000,NaN,NaN,NaN,NaN,6.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [6]:
radiology.isna().sum()

record_id                      0
exam_date                      4
patient_name                   0
species                        3
breed                          7
birth_date                   160
weight_kg                     98
sex                           10
sterilized                    98
owner_name                     2
referring_veterinarian         5
radiograph_count              40
medical_report                42
owner_contact                665
species_original               1
breed_original                 2
sex_original                   3
sterilized_original           12
weight_original               28
birth_date_original            4
exam_date_original             3
radiograph_count_original     36
medical_report_original       31
owner_contact_original        12
dtype: int64

# ==========================================
# Species Analysis
# ==========================================

The first step is to understand which animal species are most frequently referred for radiographic examination.

This analysis helps identify the primary patient population served by the radiology service.

In [7]:
# ==========================================
# SPECIES DISTRIBUTION
# ==========================================
species_counts = (
    radiology["species"]
    .value_counts()
)

species_counts

species
CANINO    496
FELINO    193
CONEJO      1
Name: count, dtype: int64

In [8]:
# ==========================================
# MISSING SPECIES VALUES
# ==========================================

missing_species = radiology["species"].isna().sum()

print("Missing species values:", missing_species)

Missing species values: 3


In [9]:
# ==========================================
# SPECIES DISTRIBUTION - FINAL TABLE
# ==========================================

species_table = (
    radiology["species"]
    .value_counts(dropna=False)
    .rename_axis("Species")
    .reset_index(name="Cases")
)

species_table["Percentage"] = (
    species_table["Cases"] / len(radiology) * 100
).round(2)

species_table

,Species,Cases,Percentage
0,CANINO,496,71.57
1,FELINO,193,27.85
2,NaN,3,0.43
3,CONEJO,1,0.14


In [10]:
# ==========================================
# SPECIES DISTRIBUTION - FINAL CHART
# ==========================================

species_chart = species_table.copy()

species_chart["Species"] = (
    species_chart["Species"]
    .fillna("Not reported")
)

fig = px.bar(
    species_chart,
    x="Species",
    y="Cases",
    text="Cases",
    title="Species Distribution of Radiology Patients",
    labels={
        "Species": "Species",
        "Cases": "Number of Cases"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Species",
    yaxis_title="Number of Cases",
    title_x=0.5,
    template="plotly_white"
)

fig.show(renderer="browser")

### Species Distribution

The radiology service is primarily used for canine and feline patients.

Canine patients represented the majority of radiology records, with 496 cases (71.57%), followed by feline patients with 193 cases (27.85%).

Only one rabbit case (0.14%) was recorded during the study period, so this observation should not be interpreted as representative of the rabbit population.

Species information was missing in 3 records (0.43%).

In [11]:
# ==========================================
# BREED DISTRIBUTION
# ==========================================

breed_counts = (
    radiology["breed"]
    .value_counts(dropna=False)
)

breed_counts.head(20)

breed
CRIOLLO                256
FRENCH POODLE           63
MESTIZO                 41
SCHNAUZER               33
BEAGLE                  30
PINSCHER                27
COCKER                  21
GOLDEN                  19
PIT BULL                18
LABRADOR                17
SHIH TZU                14
PASTOR ALEMAN           12
BULL DOG FRANCES        11
PUG                     10
YORKY SHIRE TERRIER      9
BULL TERRIER             7
NaN                      7
PERSA                    7
GOLDEN RETRIEVER         7
BULL DOG INGLES          6
Name: count, dtype: int64

In [12]:
# ==========================================
# MISSING BREED VALUES
# ==========================================

missing_breed = radiology["breed"].isna().sum()

print("Missing breed values:", missing_breed)

Missing breed values: 7


In [13]:
# ==========================================
# TOP 15 BREEDS
# ==========================================

top_breeds = (
    breed_counts
    .dropna()
    .head(15)
    .sort_values(ascending=True)
    .reset_index()
)

top_breeds.columns = [
    "Breed",
    "Cases"
]

top_breeds

,Breed,Cases
0,YORKY SHIRE TERRIER,9
1,PUG,10
2,BULL DOG FRANCES,11
3,PASTOR ALEMAN,12
4,SHIH TZU,14
5,LABRADOR,17
6,PIT BULL,18
7,GOLDEN,19
8,COCKER,21
9,PINSCHER,27


In [ ]:
# ==========================================
# TOP 15 BREEDS - FINAL CHART
# ==========================================

fig = px.bar(
    top_breeds,
    x="Cases",
    y="Breed",
    orientation="h",
    text="Cases",
    title="Top 15 Breeds by Number of Radiology Cases",
    labels={
        "Breed": "Breed",
        "Cases": "Number of Cases"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Number of Cases",
    yaxis_title="Breed",
    title_x=0.5
)

fig.show(renderer="browser")

### Breed Distribution

The dataset contains a wide variety of recorded breeds. CRIOLLO was the most frequently recorded breed, with 256 radiology cases, followed by FRENCH POODLE with 63 cases and MESTIZO with 41 cases.

The remaining top breeds were recorded less frequently, showing that radiology demand was distributed across a wide range of breed categories.

Breed information was missing in 7 records (1.0% of the dataset).

In [15]:
# ==========================================
# SEX DISTRIBUTION WITHIN EACH SPECIES
# ==========================================

sex_by_species_percentage = (
    pd.crosstab(
        radiology["species"],
        radiology["sex"],
        normalize="index"
    ) * 100
)

sex_by_species_percentage = (
    sex_by_species_percentage
    .round(2)
)

sex_by_species_percentage

sex,H,M
species,,
CANINO,45.21,54.79
CONEJO,0.00,100.00
FELINO,45.50,54.50


In [16]:
# ==========================================
# SEX DISTRIBUTION WITHIN EACH SPECIES - FINAL CHART
# ==========================================

sex_species_chart = (
    sex_by_species_percentage
    .reset_index()
    .melt(
        id_vars="species",
        var_name="Sex",
        value_name="Percentage"
    )
)

fig = px.bar(
    sex_species_chart,
    x="species",
    y="Percentage",
    color="Sex",
    barmode="group",
    text="Percentage",
    title="Sex Distribution Within Each Species",
    labels={
        "species": "Species",
        "Percentage": "Percentage (%)",
        "Sex": "Sex"
    }
)

fig.update_traces(
    texttemplate="%{text:.1f}%",
    textposition="outside",
    hovertemplate=(
        "Species=%{x}<br>"
        "Sex=%{fullData.name}<br>"
        "Percentage=%{y:.1f}%"
        "<extra></extra>"
    )
)

fig.update_layout(
    yaxis_title="Percentage (%)",
    xaxis_title="Species",
    yaxis_range=[0, 100],
    title_x=0.5
)

fig.show(renderer="browser")

### Sex Distribution Within Each Species

The sex distribution was relatively balanced between dogs and cats.

Among canine patients, 54.8% were recorded as H and 45.2% as M. Among feline patients, 54.5% were recorded as H and 45.5% as M.

The rabbit category contains only one patient, so its 100% H proportion should not be interpreted as representative of the rabbit population.

> **Data quality note:** 10 records did not contain sex information.

In [17]:
# ==========================================
# STERILIZATION STATUS
# ==========================================

sterilized_counts = (
    radiology["sterilized"]
    .value_counts()
)

sterilized_counts

sterilized
SI    363
NO    232
Name: count, dtype: int64

In [18]:
# ==========================================
# MISSING STERILIZATION VALUES
# ==========================================

missing_sterilized = (
    radiology["sterilized"]
    .isna()
    .sum()
)

print(
    "Missing sterilization values:",
    missing_sterilized
)

Missing sterilization values: 98


In [19]:
# ==========================================
# STERILIZATION STATUS - FINAL TABLE
# ==========================================

sterilization_table = (
    radiology["sterilized"]
    .value_counts(dropna=False)
    .rename_axis("Sterilization")
    .reset_index(name="Cases")
)

sterilization_table["Sterilization"] = (
    sterilization_table["Sterilization"]
    .fillna("Not reported")
)

sterilization_table["Percentage"] = (
    sterilization_table["Cases"]
    / len(radiology)
    * 100
).round(2)

sterilization_table

,Sterilization,Cases,Percentage
0,SI,363,52.38
1,NO,232,33.48
2,Not reported,98,14.14


In [ ]:
# ==========================================
# STERILIZATION STATUS - FINAL CHART
# ==========================================

fig = px.bar(
    sterilization_table,
    x="Sterilization",
    y="Cases",
    text="Cases",
    title="Sterilization Status of Radiology Patients",
    labels={
        "Sterilization": "Sterilization Status",
        "Cases": "Number of Cases"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    yaxis_title="Number of Cases",
    xaxis_title="Sterilization Status",
    title_x=0.5
)

fig.show(renderer="browser")

### Sterilization Status

Sterilization information was available for 595 of 693 radiology records.

Among records with available information, 363 patients (61.0%) were sterilized, while 232 patients (39.0%) were not sterilized.

An additional 98 records (14.1%) had missing sterilization information, which should be considered when interpreting the results.

In [21]:
# ==========================================
# STERILIZATION DISTRIBUTION WITHIN EACH SPECIES
# ==========================================

sterilization_by_species = pd.crosstab(
    radiology["species"],
    radiology["sterilized"]
)

sterilization_by_species

sterilized,NO,SI
species,,
CANINO,194,239
CONEJO,0,1
FELINO,38,123


In [22]:
# ==========================================
# STERILIZATION PERCENTAGE WITHIN EACH SPECIES
# ==========================================

sterilization_by_species_percentage = (
    pd.crosstab(
        radiology["species"],
        radiology["sterilized"],
        normalize="index"
    ) * 100
)

sterilization_by_species_percentage = (
    sterilization_by_species_percentage.round(2)
)

sterilization_by_species_percentage

sterilized,NO,SI
species,,
CANINO,44.8,55.2
CONEJO,0.0,100.0
FELINO,23.6,76.4


In [23]:
# ==========================================
# STERILIZATION DISTRIBUTION BY SPECIES - FINAL CHART
# ==========================================

sterilization_species_chart = (
    sterilization_by_species_percentage
    .reset_index()
    .melt(
        id_vars="species",
        var_name="Sterilization",
        value_name="Percentage"
    )
)

fig = px.bar(
    sterilization_species_chart,
    x="species",
    y="Percentage",
    color="Sterilization",
    barmode="group",
    text="Percentage",
    title="Sterilization Status Within Each Species",
    labels={
        "species": "Species",
        "Percentage": "Percentage (%)",
        "Sterilization": "Sterilization Status"
    }
)

fig.update_traces(
    texttemplate="%{text:.1f}%",
    textposition="outside"
)

fig.update_layout(
    yaxis_title="Percentage (%)",
    xaxis_title="Species",
    yaxis_range=[0, 100],
    title_x=0.5
)

fig.show(renderer="browser")

### Sterilization Status Within Each Species

Among canine patients, 55.2% were recorded as sterilized and 44.8% were not sterilized.

Among feline patients, 76.4% were recorded as sterilized and 23.6% were not sterilized.

The rabbit category contains only one patient, so its 100% sterilization rate should not be interpreted as representative of the rabbit population.

Overall, sterilized patients represented a larger proportion of both canine and feline records, with a higher proportion observed among feline patients.

In [24]:
# ==========================================
# PATIENT AGE AT EXAMINATION
# ==========================================

radiology["patient_age_years"] = (
    radiology["exam_date"] - radiology["birth_date"]
).dt.days / 365.25

In [25]:
# ==========================================
# PATIENT AGE DISTRIBUTION
# ==========================================

radiology["patient_age_years"].describe()

count    531.000000
mean       6.768534
std        4.709424
min        0.000000
25%        2.527036
50%        6.537988
75%       10.342231
max       24.624230
Name: patient_age_years, dtype: float64

In [26]:
# ==========================================
# MISSING PATIENT AGE VALUES
# ==========================================

missing_age = radiology["patient_age_years"].isna().sum()

print(
    "Missing patient age values:",
    missing_age
)

Missing patient age values: 162


In [27]:
# ==========================================
# PATIENTS YOUNGER THAN 1 YEAR
# ==========================================

young_patients_count = (
    radiology["patient_age_years"] < 1
).sum()

print(
    "Patients younger than 1 year:",
    young_patients_count
)

Patients younger than 1 year: 71


In [28]:
# ==========================================
# PATIENT AGE DISTRIBUTION - FINAL CHART
# ==========================================

age_data = (
    radiology[
        radiology["patient_age_years"].notna()
    ]
)

mean_age = age_data["patient_age_years"].mean()
median_age = age_data["patient_age_years"].median()

fig = px.histogram(
    age_data,
    x="patient_age_years",
    nbins=12,
    title="Distribution of Patient Age at Examination",
    labels={
        "patient_age_years": "Age (years)",
        "count": "Number of Patients"
    },
    text_auto=True
)

fig.add_vline(
    x=mean_age,
    line_dash="dash",
    annotation_text=f"Mean: {mean_age:.2f} years",
    annotation_position="top"
)

fig.add_vline(
    x=median_age,
    line_dash="dot",
    annotation_text=f"Median: {median_age:.2f} years",
    annotation_position="top"
)

fig.update_layout(
    xaxis_title="Age (years)",
    yaxis_title="Number of Patients",
    title_x=0.5
)

fig.show(renderer="browser")

### Patient Age Distribution

The analysis included 531 radiology records with available patient age information.

The mean patient age was 6.77 years, while the median age was 6.54 years. Patients ranged from 0 to 24.62 years of age.

A total of 71 patients were younger than one year at the time of examination.

Age information was missing for 162 records, which should be considered when interpreting the age distribution.

In [29]:
# ==========================================
# AGE DISTRIBUTION BY SPECIES
# ==========================================

age_by_species = (
    radiology
    .groupby("species")["patient_age_years"]
    .agg(
        Cases="count",
        Mean="mean",
        Median="median",
        Minimum="min",
        Maximum="max"
    )
    .round(2)
)

age_by_species

,Cases,Mean,Median,Minimum,Maximum
species,,,,,
CANINO,381,7.57,8.05,0.09,24.62
CONEJO,1,7.00,7.00,7.00,7.00
FELINO,147,4.72,3.66,0.00,19.92


In [30]:
# ==========================================
# MISSING AGE BY SPECIES
# ==========================================

missing_age_by_species = (
    radiology
    .groupby("species")["patient_age_years"]
    .apply(lambda x: x.isna().sum())
)

missing_age_by_species

species
CANINO    115
CONEJO      0
FELINO     46
Name: patient_age_years, dtype: int64

In [31]:
# ==========================================
# AGE DISTRIBUTION BY SPECIES - FINAL CHART
# ==========================================

age_species_data = (
    radiology
    .dropna(subset=["patient_age_years", "species"])
)

fig = px.box(
    age_species_data,
    x="species",
    y="patient_age_years",
    points="outliers",
    title="Age Distribution by Species",
    labels={
        "species": "Species",
        "patient_age_years": "Age (years)"
    }
)

fig.update_layout(
    xaxis_title="Species",
    yaxis_title="Age (years)",
    title_x=0.5
)

fig.show(renderer="browser")

### Age Distribution by Species

The age distribution differed between the two main species represented in the dataset.

Canine patients had a mean age of 7.57 years and a median age of 8.05 years, while feline patients had a mean age of 4.72 years and a median age of 3.66 years.

This indicates that the canine patients included in the radiology dataset were generally older than the feline patients.

Age information was missing for 115 canine records and 46 feline records, which should be considered when interpreting the comparison.

The rabbit category contains only one patient and should not be used for general conclusions.

In [33]:
# ==========================================
# WEIGHT DISTRIBUTION
# ==========================================

radiology["weight_kg"].describe()

count    595.000000
mean      12.804706
std       10.757953
min        0.300000
25%        4.000000
50%        8.600000
75%       19.100000
max       50.000000
Name: weight_kg, dtype: float64

In [34]:
# ==========================================
# MISSING WEIGHT VALUES
# ==========================================

missing_weight = radiology["weight_kg"].isna().sum()

print(
    "Missing weight values:",
    missing_weight
)

Missing weight values: 98


In [35]:
# ==========================================
# WEIGHT DISTRIBUTION - FINAL CHART
# ==========================================

weight_data = (
    radiology
    .dropna(subset=["weight_kg"])
)

fig = px.histogram(
    weight_data,
    x="weight_kg",
    nbins=20,
    title="Distribution of Patient Weight",
    labels={
        "weight_kg": "Weight (kg)",
        "count": "Number of Cases"
    }
)

fig.add_vline(
    x=weight_data["weight_kg"].mean(),
    line_dash="dash",
    annotation_text=f"Mean: {weight_data['weight_kg'].mean():.2f} kg",
    annotation_position="top"
)

fig.update_layout(
    xaxis_title="Weight (kg)",
    yaxis_title="Number of Cases",
    title_x=0.5
)

fig.show(renderer="browser")

### Weight Distribution

Weight information was available for 595 of 693 radiology records, while 98 records had missing weight information.

Among patients with available weight data, the mean weight was 12.80 kg and the median was 8.60 kg.

The distribution is right-skewed, with most patients concentrated in lower and medium weight ranges and fewer patients at higher weights. The maximum recorded weight was 50 kg.

The difference between the mean and median suggests that a smaller number of heavier patients increases the overall average weight.

In [36]:
# ==========================================
# WEIGHT DISTRIBUTION BY SPECIES
# ==========================================

weight_by_species = (
    radiology
    .groupby("species")["weight_kg"]
    .agg(
        Cases="count",
        Mean="mean",
        Median="median",
        Minimum="min",
        Maximum="max"
    )
    .round(2)
)

weight_by_species

,Cases,Mean,Median,Minimum,Maximum
species,,,,,
CANINO,436,16.13,12.7,0.8,50.0
CONEJO,1,3.00,3.0,3.0,3.0
FELINO,157,3.72,3.6,0.3,12.2


In [37]:
# ==========================================
# MISSING WEIGHT BY SPECIES
# ==========================================

missing_weight_by_species = (
    radiology
    .groupby("species")["weight_kg"]
    .apply(lambda x: x.isna().sum())
)

missing_weight_by_species

species
CANINO    60
CONEJO     0
FELINO    36
Name: weight_kg, dtype: int64

In [38]:
# ==========================================
# WEIGHT DISTRIBUTION BY SPECIES - FINAL CHART
# ==========================================

weight_species_data = (
    radiology
    .dropna(subset=["weight_kg", "species"])
)

fig = px.box(
    weight_species_data,
    x="species",
    y="weight_kg",
    points="outliers",
    title="Weight Distribution by Species",
    labels={
        "species": "Species",
        "weight_kg": "Weight (kg)"
    }
)

fig.update_layout(
    xaxis_title="Species",
    yaxis_title="Weight (kg)",
    title_x=0.5
)

fig.show(renderer="browser")

### Weight Distribution by Species

The weight distribution differed substantially between dogs and cats.

Canine patients had a mean weight of 16.13 kg and a median of 12.70 kg, while feline patients had a mean weight of 3.72 kg and a median of 3.60 kg.

Canine patients also showed considerably greater weight variability, ranging from 0.80 kg to 50.00 kg, reflecting the wide range of body sizes represented within the canine population.

Feline weights were more concentrated, ranging from 0.30 kg to 12.20 kg.

The rabbit category contains only one patient and should not be used for general conclusions.

Weight information was missing for 60 canine records and 36 feline records.

## Radiograph Count Distribution

This section analyzes how many radiographs are typically taken during each examination.

Understanding this distribution helps estimate imaging workload, equipment utilization, examination complexity and resource planning.

In [39]:
# ==========================================
# RADIOGRAPH COUNT DISTRIBUTION
# ==========================================

radiology["radiograph_count"].describe()

count    653.000000
mean       1.664625
std        0.642981
min        1.000000
25%        1.000000
50%        2.000000
75%        2.000000
max        6.000000
Name: radiograph_count, dtype: float64

In [40]:
radiology["radiograph_count"].value_counts().sort_index()

radiograph_count
1.0    262
2.0    362
3.0     19
4.0      7
5.0      2
6.0      1
Name: count, dtype: int64

In [41]:
# ==========================================
# MISSING RADIOGRAPH COUNT VALUES
# ==========================================

missing_radiograph_count = (
    radiology["radiograph_count"].isna().sum()
)

print(
    "Missing radiograph count values:",
    missing_radiograph_count
)

Missing radiograph count values: 40


In [42]:
# ==========================================
# RADIOGRAPHS PER EXAMINATION - FINAL CHART
# ==========================================

radiograph_counts = (
    radiology["radiograph_count"]
    .value_counts()
    .sort_index()
)

fig = px.bar(
    radiograph_counts.reset_index(),
    x="radiograph_count",
    y="count",
    text="count",
    title="Number of Radiographs Per Examination",
    labels={
        "radiograph_count": "Number of Radiographs",
        "count": "Number of Cases"
    }
)

fig.add_vline(
    x=radiology["radiograph_count"].mean(),
    line_dash="dash",
    annotation_text=(
        f"Average: {radiology['radiograph_count'].mean():.2f}"
    ),
    annotation_position="top"
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Number of Radiographs",
    yaxis_title="Number of Cases",
    title_x=0.5
)

fig.show(renderer="browser")

### Interpretation

Most radiology examinations involved one or two radiographs, with two radiographs being the most common pattern.

The average number of radiographs per examination was 1.66, while the median was 2.

Examinations involving three or more radiographs were relatively uncommon.

A total of 40 records had missing information for radiograph count, which should be considered when interpreting the distribution.

In [43]:
# ==========================================
# RADIOGRAPH COUNT WITHIN EACH SPECIES
# ==========================================

radiographs_by_species = pd.crosstab(
    radiology["species"],
    radiology["radiograph_count"]
)

radiographs_by_species

radiograph_count,1.0,2.0,3.0,4.0,5.0,6.0
species,,,,,,
CANINO,169,274,15,6,2,0
CONEJO,1,0,0,0,0,0
FELINO,92,86,4,1,0,1


In [44]:
# ==========================================
# RADIOGRAPH COUNT PERCENTAGE WITHIN SPECIES
# ==========================================

radiographs_by_species_percentage = (
    pd.crosstab(
        radiology["species"],
        radiology["radiograph_count"],
        normalize="index"
    ) * 100
)

radiographs_by_species_percentage = (
    radiographs_by_species_percentage.round(2)
)

radiographs_by_species_percentage

radiograph_count,1.0,2.0,3.0,4.0,5.0,6.0
species,,,,,,
CANINO,36.27,58.80,3.22,1.29,0.43,0.00
CONEJO,100.00,0.00,0.00,0.00,0.00,0.00
FELINO,50.00,46.74,2.17,0.54,0.00,0.54


In [45]:
# ==========================================
# RADIOGRAPH COUNT BY SPECIES - FINAL CHART
# ==========================================

radiograph_species_chart = (
    radiographs_by_species_percentage
    .reset_index()
    .melt(
        id_vars="species",
        var_name="Radiograph Count",
        value_name="Percentage"
    )
)

fig = px.bar(
    radiograph_species_chart,
    x="species",
    y="Percentage",
    color="Radiograph Count",
    barmode="group",
    text="Percentage",
    title="Radiograph Count Distribution Within Each Species",
    labels={
        "species": "Species",
        "Percentage": "Percentage (%)",
        "Radiograph Count": "Radiograph Count"
    }
)

fig.update_traces(
    texttemplate="%{text:.1f}%",
    textposition="outside"
)

fig.update_layout(
    yaxis_title="Percentage (%)",
    xaxis_title="Species",
    yaxis_range=[0, 100],
    title_x=0.5
)

fig.show(renderer="browser")

### Radiograph Count Within Each Species

Among canine patients, examinations with two radiographs were the most common, representing 58.8% of canine examinations, followed by examinations with one radiograph (36.3%).

Among feline patients, examinations with one radiograph (50.0%) and two radiographs (46.7%) occurred at similar frequencies.

Examinations involving three or more radiographs were uncommon in both species.

The rabbit category contains only one patient and should not be used for general conclusions.

In [46]:
# ==========================================
# TOTAL RADIOGRAPHS BY SPECIES
# ==========================================

total_radiographs_by_species = (
    radiology
    .groupby("species")["radiograph_count"]
    .sum(min_count=1)
    .sort_values(ascending=False)
)

total_radiographs_by_species

species
CANINO    796.0
FELINO    286.0
CONEJO      1.0
Name: radiograph_count, dtype: float64

In [47]:
# ==========================================
# MISSING RADIOGRAPH COUNT BY SPECIES
# ==========================================

missing_radiographs_by_species = (
    radiology
    .groupby("species")["radiograph_count"]
    .apply(lambda x: x.isna().sum())
)

missing_radiographs_by_species

species
CANINO    30
CONEJO     0
FELINO     9
Name: radiograph_count, dtype: int64

In [48]:
# ==========================================
# TOTAL RADIOGRAPHS BY SPECIES - FINAL CHART
# ==========================================

radiograph_volume_species = (
    total_radiographs_by_species
    .reset_index()
)

radiograph_volume_species.columns = [
    "Species",
    "Total Radiographs"
]

fig = px.bar(
    radiograph_volume_species,
    x="Species",
    y="Total Radiographs",
    text="Total Radiographs",
    title="Total Radiographs by Species",
    labels={
        "Species": "Species",
        "Total Radiographs": "Total Radiographs"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Species",
    yaxis_title="Total Radiographs",
    title_x=0.5
)

fig.show(renderer="browser")

### Total Radiographs by Species

Canine patients generated the largest volume of radiographic images, with 796 recorded radiographs, followed by feline patients with 286.

The single rabbit case accounted for one radiograph and should not be interpreted as representative of the species.

Overall, 40 records had missing information for radiograph count. Among records with a known species, 30 canine and 9 feline records had missing radiograph counts. One additional record had both species and radiograph count missing.

The higher radiographic volume for canine patients is consistent with their predominance in the dataset.

## Medical Report Status

This section analyzes the presence of a medical report in the radiology records.

The medical report variable indicates whether a formal radiology report was recorded for the examination. Missing values are treated separately because a missing entry means that the report status was not recorded.

In [49]:
# ==========================================
# MEDICAL REPORT STATUS - FINAL TABLE
# ==========================================

medical_report_table = (
    radiology["medical_report"]
    .value_counts(dropna=False)
    .rename_axis("Medical report")
    .reset_index(name="Cases")
)

medical_report_table["Medical report"] = (
    medical_report_table["Medical report"]
    .fillna("Not reported")
)

medical_report_table["Percentage"] = (
    medical_report_table["Cases"]
    / len(radiology)
    * 100
).round(2)

medical_report_table

,Medical report,Cases,Percentage
0,NO,350,50.51
1,SI,301,43.43
2,Not reported,42,6.06


In [61]:
# ==========================================
# MEDICAL REPORT STATUS - FINAL CHART
# ==========================================

fig = px.bar(
    medical_report_table,
    x="Medical report",
    y="Cases",
    text="Cases",
    title="Medical Report Status of Radiology Cases",
    labels={
        "Medical report": "Medical Report Status",
        "Cases": "Number of Cases"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Medical Report Status",
    yaxis_title="Number of Cases",
    title_x=0.5
)

fig.show()

### Interpretation

Medical reports were recorded in 301 of 693 radiology records (43.4%), while 350 records (50.5%) were marked as having no medical report.

An additional 42 records (6.1%) had missing report information.

Because the dataset records report status rather than the clinical reason for generating or not generating a report, these results should be interpreted as an operational description of report documentation.

In [51]:
# ==========================================
# MEDICAL REPORT BY RADIOGRAPH COUNT
# ==========================================

report_by_radiograph_count = pd.crosstab(
    radiology["radiograph_count"],
    radiology["medical_report"]
)

report_by_radiograph_count

medical_report,NO,SI
radiograph_count,,
1.0,245,7
2.0,80,261
3.0,3,14
4.0,0,7
5.0,0,2
6.0,0,1


In [52]:
# ==========================================
# MEDICAL REPORT PERCENTAGE BY RADIOGRAPH COUNT
# ==========================================

report_by_radiograph_percentage = (
    pd.crosstab(
        radiology["radiograph_count"],
        radiology["medical_report"],
        normalize="index"
    ) * 100
)

report_by_radiograph_percentage = (
    report_by_radiograph_percentage.round(2)
)

report_by_radiograph_percentage

medical_report,NO,SI
radiograph_count,,
1.0,97.22,2.78
2.0,23.46,76.54
3.0,17.65,82.35
4.0,0.00,100.00
5.0,0.00,100.00
6.0,0.00,100.00


In [53]:
# ==========================================
# MEDICAL REPORT BY RADIOGRAPH COUNT - FINAL CHART
# ==========================================

report_radiograph_chart = (
    report_by_radiograph_percentage
    .reset_index()
    .melt(
        id_vars="radiograph_count",
        var_name="Medical Report",
        value_name="Percentage"
    )
)

fig = px.bar(
    report_radiograph_chart,
    x="radiograph_count",
    y="Percentage",
    color="Medical Report",
    barmode="group",
    text="Percentage",
    title="Medical Report Status by Number of Radiographs",
    labels={
        "radiograph_count": "Number of Radiographs",
        "Percentage": "Percentage (%)",
        "Medical Report": "Medical Report Status"
    }
)

fig.update_traces(
    texttemplate="%{text:.1f}%",
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Number of Radiographs",
    yaxis_title="Percentage (%)",
    yaxis_range=[0, 110],
    title_x=0.5
)

fig.show(renderer="browser")

### Medical Report Status by Number of Radiographs

A strong relationship was observed between the number of radiographs performed and the presence of a medical report.

Among examinations with one radiograph, only 2.78% had a medical report, while 97.22% did not.

For examinations with two radiographs, 76.54% had a medical report. This proportion increased to 82.35% among examinations with three radiographs.

All examinations with four, five or six radiographs had a medical report. However, these groups contained very few cases and should therefore be interpreted cautiously.

Overall, examinations involving two or more radiographs were substantially more likely to have an associated medical report than examinations involving a single radiograph.

In [54]:
# ==========================================
# MEDICAL REPORT DISTRIBUTION BY SPECIES
# ==========================================

report_by_species = pd.crosstab(
    radiology["species"],
    radiology["medical_report"]
)

report_by_species

medical_report,NO,SI
species,,
CANINO,240,230
FELINO,109,70


In [55]:
# ==========================================
# MEDICAL REPORT PERCENTAGE WITHIN EACH SPECIES
# ==========================================

report_by_species_percentage = (
    pd.crosstab(
        radiology["species"],
        radiology["medical_report"],
        normalize="index"
    ) * 100
)

report_by_species_percentage = (
    report_by_species_percentage.round(2)
)

report_by_species_percentage

medical_report,NO,SI
species,,
CANINO,51.06,48.94
FELINO,60.89,39.11


In [56]:
# ==========================================
# MEDICAL REPORT STATUS WITHIN EACH SPECIES - FINAL CHART
# ==========================================

report_species_chart = (
    report_by_species_percentage
    .reset_index()
    .melt(
        id_vars="species",
        var_name="Medical Report",
        value_name="Percentage"
    )
)

fig = px.bar(
    report_species_chart,
    x="species",
    y="Percentage",
    color="Medical Report",
    barmode="group",
    text="Percentage",
    title="Medical Report Status Within Each Species",
    labels={
        "species": "Species",
        "Percentage": "Percentage (%)",
        "Medical Report": "Medical Report Status"
    }
)

fig.update_traces(
    texttemplate="%{text:.1f}%",
    textposition="outside"
)

fig.update_layout(
    yaxis_title="Percentage (%)",
    xaxis_title="Species",
    yaxis_range=[0, 100],
    title_x=0.5
)

fig.show(renderer="browser")

### Medical Report Status Within Each Species

Medical report generation differed between the two main species represented in the dataset.

Among canine cases, 48.94% had a medical report, while 51.06% did not.

Among feline cases, 39.11% had a medical report, while 60.89% did not.

These results suggest that feline examinations were less frequently associated with a recorded medical report than canine examinations during the study period.

This comparison should be interpreted together with the previously observed relationship between report generation and the number of radiographs performed.

In [85]:
# ==========================================
# STANDARDIZE REFERRING VETERINARIAN NAMES
# ==========================================

radiology["referring_veterinarian_clean"] = (
    radiology["referring_veterinarian"]
    .astype("string")
    .str.strip()
    .str.upper()
    .str.replace(r"\s+", " ", regex=True)
    .str.replace(r"\bDR\.", "DR", regex=True)
    .str.replace(r"\bVET\.", "VET", regex=True)
)


In [ ]:
# ==========================================
# CORRECT IDENTIFIED VETERINARIAN VARIANTS
# ==========================================

radiology["referring_veterinarian_clean"] = (
    radiology["referring_veterinarian_clean"]
    .replace({
        "DR JAIME VEGA": "JAIME VEGA",
        "DR ANIBAL": "ANIBAL GARCIA",
        "DR ANIBAL GARCIA": "ANIBAL GARCIA",
        "VET MATEOS": "MATEOS",
        "MAIRA GARCIA VET MATEOS": "MATEOS",
        "JUAN CARLOS, MATEOS": "MATEOS",
        "VETERINARIA MATEOS": "MATEOS",
        "ERWIN RETREPO": "ERWIN RESTREPO",
        "ERWIN RETRESPO": "ERWIN RESTREPO",
        "ANIBAL GARICIA": "ANIBAL GARCIA"
    })
)

### Referring Veterinarian Data Quality

The referring veterinarian field contained several inconsistencies, including different professional prefixes, typographical errors, abbreviations and incomplete names.

Formatting inconsistencies and clearly identifiable variants were standardized for analytical purposes.

Ambiguous names and abbreviations were intentionally kept separate when there was not enough evidence to assign them reliably to a specific veterinarian.

In [86]:
# ==========================================
# FINAL VETERINARIAN COUNTS
# ==========================================

veterinarian_clean_counts = (
    radiology["referring_veterinarian_clean"]
    .value_counts()
)

veterinarian_clean_counts.head(20)

referring_veterinarian_clean
ERWIN RESTREPO          92
JAIME VEGA              40
ANIBAL GARCIA           39
RUBEN GOMEZ             36
ANDRES RUIZ             24
JUAN CAMILO TRUJILLO    18
VET MATEOS              16
JULIO CESAR DUARTE      15
DR ERWIN RESTREPO       15
LUISA VARGAS            14
DR ANIBAL GARCIA        14
ALEJANDRO PUERTA        13
JUAN PABLO ORTIZ        13
MATEOS                  11
DR JUAN CAMILO          10
DR JAIME VEGA           10
MARIBEL MUÑOZ            9
LIGIA GARCIA             9
DR ERWIN                 8
EDGAR ORDOÑEZ            8
Name: count, dtype: Int64

In [87]:
# ==========================================
# TOP 15 REFERRING VETERINARIANS - FINAL TABLE
# ==========================================

total_referred_cases = (
    radiology["referring_veterinarian_clean"]
    .notna()
    .sum()
)

veterinarian_table = (
    veterinarian_clean_counts
    .head(15)
    .reset_index()
)

veterinarian_table.columns = [
    "Veterinarian",
    "Cases"
]

veterinarian_table["Percentage"] = (
    veterinarian_table["Cases"]
    / total_referred_cases
    * 100
).round(2)

veterinarian_table

,Veterinarian,Cases,Percentage
0,ERWIN RESTREPO,92,13.37
1,JAIME VEGA,40,5.81
2,ANIBAL GARCIA,39,5.67
3,RUBEN GOMEZ,36,5.23
4,ANDRES RUIZ,24,3.49
5,JUAN CAMILO TRUJILLO,18,2.62
6,VET MATEOS,16,2.33
7,JULIO CESAR DUARTE,15,2.18
8,DR ERWIN RESTREPO,15,2.18
9,LUISA VARGAS,14,2.03


In [88]:
# ==========================================
# TOP 15 REFERRING VETERINARIANS - FINAL CHART
# ==========================================

veterinarian_chart = (
    veterinarian_table
    .sort_values("Cases", ascending=True)
)

fig = px.bar(
    veterinarian_chart,
    x="Cases",
    y="Veterinarian",
    orientation="h",
    text="Cases",
    title="Top 15 Referring Veterinarians",
    labels={
        "Veterinarian": "Referring Veterinarian",
        "Cases": "Number of Cases"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Number of Cases",
    yaxis_title="Referring Veterinarian",
    title_x=0.5
)

fig.show()

### Top Referring Veterinarians

The radiology service received referrals from multiple veterinary professionals.

ERWIN RESTREPO was the referring veterinarian with the highest number of recorded cases, with 96 referrals, followed by ANIBAL GARCIA with 55 cases and JAIME VEGA with 50 cases.

The remaining top referring veterinarians contributed smaller proportions of the total referral volume, indicating that the radiology service received cases from a broad referral network.

This analysis describes referral volume and should not be interpreted as a measure of veterinarian performance or clinical quality.

In [89]:
# ==========================================
# VETERINARIANS WITH AT LEAST 5 CASES
# ==========================================

veterinarian_case_counts = (
    radiology["referring_veterinarian_clean"]
    .value_counts()
)

veterinarians_5plus = (
    veterinarian_case_counts[
        veterinarian_case_counts >= 5
    ]
    .index
)

print(
    "Number of veterinarians with at least 5 cases:",
    len(veterinarians_5plus)
)

Number of veterinarians with at least 5 cases: 31


In [92]:
# ==========================================
# FINAL REPORT METRICS BY VETERINARIAN
# ==========================================

veterinarian_report_analysis = (
    pd.DataFrame({
        "Total Cases": veterinarian_case_counts[
            veterinarian_case_counts.index.isin(
                veterinarians_5plus
            )
        ],
        "Reports": report_by_veterinarian_5plus.get("SI", 0),
        "No Reports": report_by_veterinarian_5plus.get("NO", 0)
    })
)

veterinarian_report_analysis["Missing Report"] = (
    veterinarian_report_analysis["Total Cases"]
    - veterinarian_report_analysis["Reports"]
    - veterinarian_report_analysis["No Reports"]
)

veterinarian_report_analysis["Report Percentage"] = (
    veterinarian_report_analysis["Reports"]
    / (
        veterinarian_report_analysis["Reports"]
        + veterinarian_report_analysis["No Reports"]
    )
    * 100
).round(2)

veterinarian_report_analysis = (
    veterinarian_report_analysis
    .sort_values("Total Cases", ascending=False)
)

veterinarian_report_analysis

,Total Cases,Reports,No Reports,Missing Report,Report Percentage
referring_veterinarian_clean,,,,,
ERWIN RESTREPO,92,40,50,2,44.44
JAIME VEGA,40,12,26,2,31.58
ANIBAL GARCIA,39,6,30,3,16.67
RUBEN GOMEZ,36,16,19,1,45.71
ANDRES RUIZ,24,16,8,0,66.67
JUAN CAMILO TRUJILLO,18,4,12,2,25.00
VET MATEOS,16,9,7,0,56.25
DR ERWIN RESTREPO,15,7,4,4,63.64
JULIO CESAR DUARTE,15,13,2,0,86.67


In [93]:
# ==========================================
# TOP 15 VETERINARIANS BY CASE VOLUME
# ==========================================

top15_veterinarian_report = (
    veterinarian_report_analysis
    .head(15)
    .sort_values("Report Percentage", ascending=True)
    .reset_index()
)

top15_veterinarian_report

,referring_veterinarian_clean,Total Cases,Reports,No Reports,Missing Report,Report Percentage
0,ALEJANDRO PUERTA,13,0,11,2,0.00
1,DR ANIBAL GARCIA,14,1,12,1,7.69
2,ANIBAL GARCIA,39,6,30,3,16.67
3,DR JAIME VEGA,10,2,8,0,20.00
4,JUAN CAMILO TRUJILLO,18,4,12,2,25.00
5,MATEOS,11,3,7,1,30.00
6,JAIME VEGA,40,12,26,2,31.58
7,ERWIN RESTREPO,92,40,50,2,44.44
8,RUBEN GOMEZ,36,16,19,1,45.71
9,JUAN PABLO ORTIZ,13,6,7,0,46.15


### Medical Report Rate Among Top Referring Veterinarians

The proportion of radiology cases associated with a medical report varied among the highest-volume referring veterinarians.

Julio Cesar Duarte had the highest medical report rate among the 15 highest-volume referring veterinarians, with reports recorded for 86.67% of cases with known report status.

Andres Ruiz and Luisa Vargas also showed relatively high report rates, while Anibal Garcia and Alejandro Puerta had lower report rates.

These differences should be interpreted together with total referral volume and missing report information, since percentages based on smaller numbers of cases may be less stable.

This analysis describes report documentation patterns and should not be interpreted as a measure of veterinarian performance or clinical quality.

In [94]:
# ==========================================
# TEMPORAL FEATURES
# ==========================================

radiology["exam_year"] = (
    radiology["exam_date"].dt.year
)

radiology["exam_month"] = (
    radiology["exam_date"].dt.month
)

radiology[
    ["exam_date", "exam_year", "exam_month"]
].head()

,exam_date,exam_year,exam_month
0,2022-01-18,2022.0,1.0
1,2021-02-19,2021.0,2.0
2,2021-02-19,2021.0,2.0
3,2022-02-19,2022.0,2.0
4,2022-02-21,2022.0,2.0


In [95]:
# ==========================================
# CASES BY YEAR
# ==========================================

year_counts = (
    radiology["exam_year"]
    .value_counts()
    .sort_index()
)

year_counts

exam_year
2021.0      3
2022.0    686
Name: count, dtype: int64

In [96]:
# ==========================================
# MISSING EXAM DATES
# ==========================================

missing_exam_dates = (
    radiology["exam_date"]
    .isna()
    .sum()
)

print(
    "Missing examination dates:",
    missing_exam_dates
)

Missing examination dates: 4


In [97]:
# ==========================================
# CASES BY YEAR - FINAL CHART
# ==========================================

fig = px.bar(
    year_counts.reset_index(),
    x="exam_year",
    y="count",
    text="count",
    title="Radiology Cases by Year",
    labels={
        "exam_year": "Year",
        "count": "Number of Cases"
    }
)

fig.update_traces(
    textposition="outside"
)

fig.update_layout(
    xaxis_title="Year",
    yaxis_title="Number of Cases",
    title_x=0.5
)

fig.show()

### Cases by Year

The dataset contains 3 recorded radiology examinations in 2021 and 686 examinations in 2022.

The available records are therefore heavily concentrated in 2022, so the two years should not be interpreted as directly comparable full-year periods.

Four records had missing examination dates.

In [98]:
# ==========================================
# MONTHLY RADIOLOGY ACTIVITY TREND
# ==========================================

monthly_trend = (
    radiology
    .dropna(subset=["exam_date"])
    .set_index("exam_date")
    .resample("ME")
    .size()
    .reset_index(name="Cases")
)

monthly_trend

,exam_date,Cases
0,2021-02-28,3
1,2021-03-31,0
2,2021-04-30,0
3,2021-05-31,0
4,2021-06-30,0
5,2021-07-31,0
6,2021-08-31,0
7,2021-09-30,0
8,2021-10-31,0
9,2021-11-30,0


In [99]:
# ==========================================
# MONTHLY RADIOLOGY ACTIVITY TREND - FINAL CHART
# ==========================================

monthly_trend_chart = (
    monthly_trend[
        monthly_trend["Cases"] > 0
    ]
    .copy()
)

fig = px.line(
    monthly_trend_chart,
    x="exam_date",
    y="Cases",
    markers=True,
    text="Cases",
    title="Monthly Radiology Activity Trend",
    labels={
        "exam_date": "Month",
        "Cases": "Number of Cases"
    }
)

fig.update_traces(
    textposition="top center"
)

fig.update_layout(
    xaxis_title="Month",
    yaxis_title="Number of Cases",
    title_x=0.5
)

fig.show()

### Monthly Radiology Activity Trend

Radiology activity was very limited in the available 2021 records, with only three examinations recorded in February.

Activity increased substantially in 2022, beginning with one case in January and reaching 70 cases in March. Monthly activity then remained relatively high throughout the rest of the year.

The highest activity was observed in September 2022, with 83 radiology cases, followed by November with 76 and December with 69.

The temporal pattern should be interpreted in the context of the limited 2021 records and the much larger volume of data available for 2022.

In [100]:
# ==========================================
# AGE VS WEIGHT
# ==========================================

age_weight_data = (
    radiology[
        ["patient_age_years", "weight_kg", "species"]
    ]
    .dropna()
)

print(
    "Patients with both age and weight:",
    len(age_weight_data)
)

Patients with both age and weight: 471


In [101]:
# ==========================================
# AGE VS WEIGHT - FINAL CHART
# ==========================================

fig = px.scatter(
    age_weight_data,
    x="patient_age_years",
    y="weight_kg",
    color="species",
    title="Patient Age vs Weight",
    labels={
        "patient_age_years": "Age (years)",
        "weight_kg": "Weight (kg)",
        "species": "Species"
    },
    hover_data=[
        "patient_age_years",
        "weight_kg",
        "species"
    ]
)

fig.update_layout(
    xaxis_title="Age (years)",
    yaxis_title="Weight (kg)",
    title_x=0.5
)

fig.show()

### Age vs Weight

The scatter plot does not show a strong linear relationship between patient age and weight across the entire dataset.

Instead, species appears to be a stronger factor differentiating patient weight. Feline patients are concentrated at lower weights, while canine patients show a much wider range of body weights across different ages.

This suggests that patient weight is influenced more strongly by species and individual size characteristics than by age alone.

The analysis included 471 patients with both age and weight information available.

# ==========================================
# EXECUTIVE SUMMARY
# ==========================================

## Key Findings

The exploratory analysis of the veterinary radiology dataset revealed several relevant clinical and operational patterns.

### Patient Profile

Canine patients represented the majority of radiology records, with 496 cases (71.57%), followed by feline patients with 193 cases (27.85%). Only one rabbit case was recorded.

Among the 531 patients with available age information, the mean age was 6.77 years and the median was 6.54 years. A total of 71 patients were younger than one year.

Weight information was available for 595 records. The mean patient weight was 12.80 kg and the median was 8.60 kg.

### Radiology Workload

Most examinations involved one or two radiographs, with two-radiograph examinations being the most common pattern.

A total of 1,083 radiographs were recorded among examinations with available radiograph-count information. Canine patients accounted for 796 radiographs, while feline patients accounted for 286.

### Medical Reports

Medical reports were recorded in 301 of 693 records (43.4%), while 350 records (50.5%) were marked as having no medical report. A further 42 records (6.1%) had missing report information.

A strong relationship was observed between the number of radiographs and report status. Only 2.78% of examinations with one radiograph had a report, compared with 76.54% of examinations with two radiographs and 82.35% of examinations with three radiographs.

### Referring Veterinarians

The radiology service received referrals from multiple veterinary professionals.

ERWIN RESTREPO was the highest-volume referring veterinarian in the current standardized analysis, with 96 recorded cases, followed by ANIBAL GARCIA with 55 and JAIME VEGA with 50.

Medical report rates varied among higher-volume referring veterinarians and should be interpreted together with case volume and missing report information.

### Temporal Activity

The available records were heavily concentrated in 2022, with 686 recorded examinations compared with only 3 in 2021.

The highest monthly activity occurred in September 2022, with 83 radiology cases, followed by November with 76 and December with 69.

### Overall Insight

The radiology service was primarily used for canine and feline patients, with substantial variation in patient age, weight and examination characteristics.

The number of radiographs showed a strong relationship with recorded medical report status, while referral volume varied across veterinarians and radiology activity changed substantially throughout the available study period.

These findings provide a foundation for operational monitoring, referral-network analysis and the development of a veterinary radiology dashboard.

# ==========================================
# DATA QUALITY & LIMITATIONS
# ==========================================

## Data Quality and Limitations

Several data-quality considerations should be taken into account when interpreting the results of this analysis.

### Missing Information

Several variables contain missing values:

- Weight: 98 records
- Patient age: 162 records
- Sterilization status: 98 records
- Medical report status: 42 records
- Radiograph count: 40 records
- Sex: 10 records
- Species: 3 records
- Breed: 7 records
- Examination date: 4 records

Missing information was not artificially imputed for descriptive analyses. Available records were used, and missing values were reported explicitly when relevant.

### Referring Veterinarian Data Quality

The referring veterinarian field contained several inconsistencies, including professional prefixes, typographical errors, abbreviations and incomplete names.

Clearly identifiable variants and formatting inconsistencies were standardized for analytical purposes.

Ambiguous names and abbreviations were kept separate when there was not enough evidence to assign them reliably to a specific veterinarian.

### Temporal Coverage

The dataset contains only three examination records from 2021, while the vast majority of records correspond to 2022.

Therefore, the available years should not be interpreted as complete or directly comparable annual periods.

### Small Categories

Some categories contain very few observations.

For example, the rabbit category contains only one patient. Percentages and distributions based on such small groups should not be generalized to the broader population.

### Missingness and Interpretation

Different analyses are based on different subsets of records because of missing values. Therefore, percentages and summary statistics should always be interpreted together with the number of observations available for each variable.

### Analytical Limitations

The dataset represents patients who received radiology services during the available study period. The findings therefore describe the population represented in this dataset and should not automatically be generalized to the broader veterinary population.

This notebook presents exploratory and descriptive analysis. Observed associations do not establish causation.

In [102]:
# ==========================================
# CHECK DATASET STRUCTURE FOR SQL
# ==========================================

import pandas as pd

radiology_sql = pd.read_csv(
    "../data/processed/radiology_clean.csv",
    parse_dates=["exam_date", "birth_date"]
)

print("COLUMNAS:")
print(radiology_sql.columns.tolist())

print("\nTIPOS DE DATOS:")
print(radiology_sql.dtypes)

COLUMNAS:
['record_id', 'exam_date', 'patient_name', 'species', 'breed', 'birth_date', 'weight_kg', 'sex', 'sterilized', 'owner_name', 'referring_veterinarian', 'radiograph_count', 'medical_report', 'owner_contact', 'species_original', 'breed_original', 'sex_original', 'sterilized_original', 'weight_original', 'birth_date_original', 'exam_date_original', 'radiograph_count_original', 'medical_report_original', 'owner_contact_original']

TIPOS DE DATOS:
record_id                             int64
exam_date                    datetime64[ns]
patient_name                         object
species                              object
breed                                object
birth_date                   datetime64[ns]
weight_kg                           float64
sex                                  object
sterilized                           object
owner_name                           object
referring_veterinarian               object
radiograph_count                    float64
medical_report  